addons/token-exchange/agent_mirrorpeer_controller.go

```diff
@@ -0,0 +1,178 @@
+		return ctrl.Result{}, fmt.Errorf("failed to enable mirroring the storagecluster %q in namespace %q in managed cluster. Error %v", scr.Name, scr.Namespace, err)
+	}
+
+	err = r.createVolumeReplicationClass(ctx, &mirrorPeer)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create volumereplicationclass: %v", err)
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *MirrorPeerReconciler) getStorageClusterDetails(mp *multiclusterv1alpha1.MirrorPeer) *multiclusterv1alpha1.StorageClusterRef {
+	for _, v := range mp.Spec.Items {
+		if v.ClusterName == r.SpokeClusterName {
+			return &v.StorageClusterRef
+		}
+	}
+	return nil
+}
+
+func (r *MirrorPeerReconciler) enableMirroring(ctx context.Context, storageClusterName string, namespace string) error {
+	var sc ocsv1.StorageCluster
+	err := r.SpokeClient.Get(ctx, types.NamespacedName{
+		Name:      storageClusterName,
+		Namespace: namespace,
+	}, &sc)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return fmt.Errorf("could not find storagecluster %q in namespace %v: %v", storageClusterName, namespace, err)
+		}
+		return err
+	}
+	sc.Spec.Mirroring.Enabled = true
+	return r.SpokeClient.Update(ctx, &sc)
+}
+
+func (r *MirrorPeerReconciler) enableCSIOmapGenerator(ctx context.Context, namespace string) error {
+	var rcm corev1.ConfigMap
+	err := r.SpokeClient.Get(ctx, types.NamespacedName{
+		Name:      RookConfigMapName,
+		Namespace: namespace,
+	}, &rcm)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return fmt.Errorf("could not find rook-ceph-config-map: %v", err)
+		}
+		return err
+	}
+
+	rcm.Data[RookCSIEnableKey] = "true"
+	rcm.Data[RookVolumeRepKey] = "true"
+	return r.SpokeClient.Update(ctx, &rcm)
+}
+
+func (r *MirrorPeerReconciler) createVolumeReplicationClass(ctx context.Context, mp *multiclusterv1alpha1.MirrorPeer) error {
+	scr := r.getStorageClusterDetails(mp)
+	params := make(map[string]string)
+	params[MirroringModeKey] = string(mp.Spec.Mode)
+	params[SchedulingIntervalKey] = mp.Spec.SchedulingInterval
+	params[ReplicationSecretNameKey] = mp.Spec.ReplicationSecretName
+	params[ReplicationSecretNamespaceKey] = scr.Namespace
+	vrc := &replicationv1alpha1.VolumeReplicationClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "odf-rbd-volumereplicationclass",
+		},
+		Spec: replicationv1alpha1.VolumeReplicationClassSpec{
+			Provisioner: Provisioner,
+			Parameters:  params,
+		},
+	}
+	var found replicationv1alpha1.VolumeReplicationClass
+	err := r.SpokeClient.Get(ctx, types.NamespacedName{
+		Name:      vrc.Name,
+		Namespace: vrc.Namespace,
+	}, &found)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			err = r.SpokeClient.Create(ctx, vrc)
+			return err
+		}
+		return err
+	}
+	found.Spec = vrc.Spec
+	return r.SpokeClient.Update(ctx, &found)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *MirrorPeerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	mpPredicate := common.ComposePredicates(predicate.GenerationChangedPredicate{})
```

Comment:
can we add another predicate such that only MirrorPeer that targets this spoke cluster is reconciled instead of all?